rfcs/0000-variable-length-header-field/0000-variable-length-header-field.md

```diff
@@ -0,0 +1,154 @@
+
+This document proposes a solution to add a variable length field to the block header. How the new field is interpreted is beyond the scope of this document and must be defined and deployed via a future soft fork.
+
+## Specification
+
+The block header is encoded as a molecule struct, which consists of fixed length fields. The header binary is just the concatenation of all the fields in sequence.
+
+There are many different ways to add the variable length field to the block header. This RFC proposes to replace the `uncles_hash` in the header with the new field `extra_hash`, which is also a 64-bytes hash. The block will have a new field `extension`.
+
+The following paragraphs will refer to the current running CKB version as ckb2019, the next fork as ckb2021 and the soft fork introducing the schema of `extension` as ckb2021x.
+
+In CKB ckb2019 and blocks generated in ckb2019, `extra_hash` is the same with the `uncles_hash`. So the block headers serialized by molecule are identical. The version `ckb2019` requires that the `extra_hash` is the hash on `uncles` in the block body. It is all zeros when `uncles` is empty, or `ckbhash` on all the uncle header hashes concatenated together.
+
+    ```
+    uncles_hash = 0 when uncles is empty, otherwise
+
+    uncles_hash = ckbhash(U1 || U2 || ... || Un)
+      where Ui is the header_hash of the i-th uncle in uncles
+    ```
+
+See Appendix for the default hash function `ckbhash`. The annotation `||` means bytes concatenation.
+
+The version ckb2021 only cares about the `extension_hash`:
+
+```
+extension_hash = 0 when extension is empty, otherwise
+
+extension_hash = ckbhash(extension)
+```
+
+The `extra_hash` is defined as:
+
+* When `extension_hash` is 0, `extra_hash` is the same with the `uncles_hash`.
+* Otherwise `extra_hash = ckbhash(uncles_hash || extension_hash)`
+
+The version ckb2021x will add consensus rule on the schema of `extension` and nodes must synchronize the field to verify that it is valid.
+
+### P2P Protocols Changes
+
+The field `uncles_hash` in the block header is renamed to `extra_hash`.
+
+```
+struct RawHeader {
+    version:                Uint32,
+    compact_target:         Uint32,
+    timestamp:              Uint64,
+    number:                 Uint64,
+    epoch:                  Uint64,
+    parent_hash:            Byte32,
+    transactions_root:      Byte32,
+    proposals_hash:         Byte32,
+    extra_hash:             Byte32,
+    dao:                    Byte32,
+}
+```
+
+The new field `extension_hash` will be added to the block body and following data structures:
+
+```
+table Block {
+    header:         Header,
+    uncles:         UncleBlockVec,
+    transactions:   TransactionVec,
+    proposals:      ProposalShortIdVec,
+    extension_hash: Byte32,
+}
+
+table BlockTransactions {
+    block_hash:     Byte32,
+    transactions:   TransactionVec,
+    uncles:         UncleBlockVec,
+    extension_hash: Byte32,
```

Comment:
I think this field can be omitted, because of `CompactBlock` has this field already, peer can use that field to reconstruct the block.